File: src/gestorAplicacion/clasesPrincipales/Rutina.java
package gestorAplicacion.clasesPrincipales;
import gestorAplicacion.clasesEnum.*;
import gestorAplicacion.clasesHerencia.Movimiento;


import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

//Autor: Jacobo Ochoa 
//Finalidad de la clase: Simular una rutina de ejercicios para un entrenamiento

public class Rutina implements Serializable{
    private static final long serialVersionUID = 1L;
	
  //Componentes: atributos de nombre, TipoEjercicio y una constante con el máximo numero de ejercicios en una rutina
    //Estructura de datos: ArrayList con todos los ejercicios que hacen parte de una rutina
    
    //Cabeceras en los métodos
    //A tener en cuenta
    
    private String nombre;
	private ArrayList<Ejercicio> ejercicios = new ArrayList<Ejercicio>();
	private TipoEjercicio tipoEjercicio;
	private static final int NUMEJERCIOS = 7; //Uso de constante

    public Rutina(String nombre) {
    	this.setNombre(nombre);
    }

    //Sobrecarga de constructores
    
    public Rutina(ArrayList<Ejercicio> ejercicios, TipoEjercicio tipoEjercicio){
    	this.ejercicios = ejercicios;
    	this.tipoEjercicio = tipoEjercicio;
    }
    
    public Rutina(String nombre, ArrayList<Ejercicio> ejercicios, TipoEjercicio tipoEjercicio){
        this.nombre = nombre;
    	this.ejercicios = ejercicios;
        this.tipoEjercicio = tipoEjercicio;
    }

    //Genera un objeto de rutina con los parametros de Tipo de ejercio, objetivo y el Nivel del cliente
    //Este debe tener 1 calentamiento, hasta 5 ejercios del tipo de ejercicio y un estiramiento
    
    public static Rutina generarRutina(TipoEjercicio tipoEjercicio, ObjetivoCliente objetivo, NivelCliente dificultad){
        ArrayList<Ejercicio> ejercicios = new ArrayList<Ejercicio>();
        
        //Filtra los movimientos por dificultad y por tipo de movimiento y los agrega a un arrayList

        ArrayList<Movimiento> movimientosDisponibles = Movimiento.filtrarMovimientosPorDificultad(Movimiento.filtrarMovimientosPorTipo(tipoEjercicio), dificultad);
        Random random = new Random();

        //Filtra los movimientos por tipo para que entregue los de tipo Calentmiento y estiramiento

        ArrayList<Movimiento> movCalentamiento = Movimiento.filtrarMovimientosPorTipo(TipoEjercicio.CALENTAMIENTO);
        ArrayList<Movimiento> movEstiramiento = Movimiento.filtrarMovimientosPorTipo(TipoEjercicio.ESTIRAMIENTO);
        
        //Convierte los movimientos en ejercicios/agrega el calentamiento

        ejercicios.add(Ejercicio.generarEjercicio(movCalentamiento.get(random.nextInt(movCalentamiento.size())), objetivo ));

        Collections.shuffle(movimientosDisponibles);

        //Añade los movimientos convertidos a ejercicios que son de tipoEjercicio a la lista de ejercicios

        for (int i = 0; i < movimientosDisponibles.size(); i++){
            if (i==(NUMEJERCIOS - 2)){
                break;
            }else {
                ejercicios.add(Ejercicio.generarEjercicio(movimientosDisponibles.get(i), objetivo));
            }
        }
        //Convierte los movimientos en ejercicios/agrega el estiramiento

        ejercicios.add(Ejercicio.generarEjercicio(movEstiramiento.get(random.nextInt(movEstiramiento.size())), objetivo ));
        
        return new Rutina(ejercicios, tipoEjercicio);
    }

    //toString que retorna la rutina con el nombre del ejercicio, y cuántas series y repeticiones se deben realizar
   
    @Override
    public String toString(){
        String toString = getTipoEjercicio() + "\n";
        for (Ejercicio ejercicio : getEjercicios()){
            toString = toString 
	            		+ ejercicio.getNombre() 
	            		+ " = " 
	            		+ ejercicio.getMaquina().getNombre() 
	            		+ " ---> " 
	            		+ ejercicio.getRepeticiones() 
	            		+ "x" 
	            		+ ejercicio.getSeries() 
	            		+ "\n";
        }
        return toString;
    }

	//Metodos Set y Get
    
    public void setEjercicios (ArrayList<Ejercicio> ejercicios) {
        this.ejercicios = ejercicios;
    }
    
    public void setTipoEjercicio (TipoEjercicio tipoEjercicio) {
        this.tipoEjercicio = tipoEjercicio;
    }
    
    public void setNombre(String nombre) {
		this.nombre = nombre;
    }

    public ArrayList<Ejercicio> getEjercicios() {
    	return ejercicios;
    }
    
    public TipoEjercicio getTipoEjercicio() {
    	return tipoEjercicio;
    }
    
    public String getNombre() {
        return nombre;
    }
      
    public int retNumEje() {
    	return NUMEJERCIOS;
    }
}
    

File: Python/src/gestorAplicacion/clasesPrincipales/Rutina.py
import random
from src.gestorAplicacion.clasesEnum import TipoEjercicio, NivelCliente, ObjetivoCliente
from src.gestorAplicacion.clasesHerencia.Movimiento import Movimiento
from src.gestorAplicacion.clasesPrincipales.Ejercicio import Ejercicio

# Autor: Jacobo Ochoa
# Finalidad de la clase: Simular una rutina de ejercicios para un entrenamiento


class Rutina:
    NUMEJERCIOS = 7

    def __init__(self, nombre=None, ejercicios=None, tipoEjercicio=None):
        self.nombre = nombre
        self.ejercicios = ejercicios if ejercicios else []
        self.tipoEjercicio = tipoEjercicio

    # Componentes: atributos de nombre, TipoEjercicio y una constante con el máximo número de ejercicios en una rutina
    # Estructura de datos: Lista con todos los ejercicios que hacen parte de una rutina

    @staticmethod
    def generarRutina(tipoEjercicio: TipoEjercicio, objetivo: ObjetivoCliente, dificultad: NivelCliente):
        ejercicios = []

        # Filtra los movimientos por dificultad y por tipo de movimiento y los agrega a una lista

        movimientosDisponibles = Movimiento.filtrarMovimientosPorDificultad(Movimiento.filtrarMovimientosPorTipo(tipoEjercicio), dificultad)

        # Filtra los movimientos por tipo para obtener aquellos de tipo Calentamiento y Estiramiento

        movCalentamiento = Movimiento.filtrarMovimientosPorTipo(TipoEjercicio.Enum.CALENTAMIENTO)
        movEstiramiento = Movimiento.filtrarMovimientosPorTipo(TipoEjercicio.Enum.ESTIRAMIENTO)

        # Convierte los movimientos en ejercicios

        ejercicios.append(Ejercicio.generarEjercicio(random.choice(movCalentamiento), objetivo))

        # Añade los ejercicios a la lista para retornar la rutina

        random.shuffle(movimientosDisponibles)

        for i in range(len(movimientosDisponibles)):
            if i == (Rutina.NUMEJERCIOS - 2):
                break
            else:
                ejercicios.append(Ejercicio.generarEjercicio(movimientosDisponibles[i], objetivo))

        ejercicios.append(Ejercicio.generarEjercicio(random.choice(movEstiramiento), objetivo))
        
        return Rutina(ejercicios, tipoEjercicio)

    def __str__(self):
        # Muestra la rutina en pantalla
        toString = str(self.tipoEjercicio) + "\n"
        for ejercicio in self.ejercicios:
            toString += "{} = {} ---> {}x{}\n".format(ejercicio.nombre, ejercicio.maquina.nombre, ejercicio.repeticiones, ejercicio.series)
        return toString
